pkg/fetcher/fetcher.go

```diff
@@ -414,6 +428,56 @@ func (u *URLFetcher) IsNonretryableClientError() bool {
 		s != http.StatusLocked && s != http.StatusTooManyRequests
 }
 
+// buildRegistryErrMsg builds error message for unexpected http code (nonretryable client errors and all other errors)
+// and extracts message details from response body stream if there is one (#5951).
+func (u *URLFetcher) buildRegistryErrMsg(httpStatusCode int, url *url.URL, respBody io.ReadCloser) string {
+	errMsg := fmt.Sprintf("Unexpected http code: %d (%s), URL: %s", httpStatusCode, http.StatusText(httpStatusCode), url)
+
+	errDetail, err := u.extractErrResponseMessage(respBody)
+	if err != nil {
+		return errMsg
+	}
+
+	errMsg += fmt.Sprintf(", Message: %s", errDetail)
+	return errMsg
+}
+
+// extractErrResponseMessage extracts `message` field from error response body json (#5951).
+func (u *URLFetcher) extractErrResponseMessage(rdr io.ReadCloser) (string, error) {
+	out := bytes.NewBuffer(nil)
+	_, err := io.Copy(out, rdr)
+	if err != nil {
+		log.Debugf("Error when copying from error response body stream: %s", err)
+		return "", err
+	}
+
+	res := []byte(out.Bytes())
+	log.Debugf("Error message json string: %s", string(res))
+
+	var errResponse RegistryErrorRespBody
+	err = json.Unmarshal(res, &errResponse)
+	if err != nil {
+		log.Debugf("Error when unmarshalling error response body: %s", err)
+		return "", err
+	}
+
+	if len(errResponse.Errors) == 0 {
+		log.Debugf("Error response wrong format. Response body: %s", string(res))
+		return "", fmt.Errorf("error response json has unconventional format")
+	}
+
+	// grab out every error message
+	var errString string
+	for i, error := range errResponse.Errors {
```

Comment:
This loop can be made more efficient by avoiding a copy of each element of `errResponse.Errors` at every iteration, like so:

```go
	for i := range errResponse.Errors {
		if i > 0 {
			errString += ", "
		}
		errString += errResponse.Errors[i].Message
	}
```

For more details, read the Golang spec for `for` statements: https://golang.org/ref/spec#For_statements